jvm/core/src/main/scala/org/scalatest/prop/Generator.scala

```diff
@@ -1357,56 +1353,52 @@ object Generator {
     */
   implicit val posZFloatGenerator: Generator[PosZFloat] =
     new Generator[PosZFloat] {
+
+      case class NextRoseTree(value: PosZFloat) extends RoseTree[PosZFloat] {
+        def shrinks(rndPassedToShrinks: Randomizer): (List[RoseTree[PosZFloat]], Randomizer) = {
+          @tailrec
+          def shrinkLoop(f: PosZFloat, acc: List[RoseTree[PosZFloat]]): List[RoseTree[PosZFloat]] = {
+            val fv = f.value
+            if (fv == 0.0f) acc
+            else if (fv <= 1.0f) NextRoseTree(PosZFloat(0.0f)) :: acc
+            else if (!fv.isWhole) {
+              val n =
+                if (fv == Float.PositiveInfinity || fv.isNaN)
+                  Float.MaxValue
+                else fv
+              // Nearest whole numbers closer to zero
+              val nearest = PosZFloat.ensuringValid(n.floor)
+              shrinkLoop(nearest, NextRoseTree(nearest) :: acc)
+            }
+            else {
+              val sqrt: Float = math.sqrt(fv.toDouble).toFloat
+              if (sqrt < 1.0f) NextRoseTree(PosZFloat(0.0f)) :: acc
+              else {
```

Comment:
@cheeseng I think if if we are not recursively calling shrinkLoop we should probably use Rose, because we are at the end of the line. I'm not 100% sure, but let's stick with that pattern unless we get evidence to the contrary. We need to make sure that shrinking terminates, so at some point there needs to be a node that has no subnodes. That's what the Rose is. I'll make such changes in this PR before merging.

Reply:
@bvenners Noted, I'll make the related changes and include them in the next PR.

Thanks.